File: lab_12_langgraph_starter/agents/fetch-users-agent.js
/**
 * Fetch Users Agent – has a tool to call the Random User API.
 * Fetches N random users from https://randomuser.me/api/
 * Used as a node in the LangGraph; reads userCount from state, writes randomUsers.
 */

const RANDOM_USER_API = "https://randomuser.me/api/";

/**
 * Tool: call Random User API and return the results array.
 * @param {number} count - Number of users to fetch (1–5000 per API docs).
 * @returns {Promise<Object[]>} Array of user objects from the API.
 */
export async function randomUsersApi(count) {
  const n = Math.min(Math.max(1, Number(count) || 10), 100);
  const url = `${RANDOM_USER_API}?results=${n}`;
  const res = await fetch(url);
  if (!res.ok) throw new Error(`Random User API error: ${res.status}`);
  const data = await res.json();
  return data.results ?? [];
}

export class FetchUsersAgent {
  constructor() {
    this.name = "FetchUsersAgent";
  }

  /**
   * Node function: uses randomUsersApi tool to fetch users, returns partial state update.
   * @param {Object} state - Graph state with userCount
   * @returns {Promise<{ randomUsers: Object[] }>}
   */
  async run(state) {
    const userCount = state.userCount ?? 10;
    const users = await randomUsersApi(userCount);
    return { randomUsers: users };
  }

  getInfo() {
    return { name: this.name, role: "Fetch random users via Random User API" };
  }
}

File: lab_12_langgraph_starter_py/agents/fetch_users_agent.py
"""
Fetch Users Agent – calls the Random User API.
Fetches N random users from https://randomuser.me/api/
"""
import requests

RANDOM_USER_API = "https://randomuser.me/api/"


def random_users_api(count: int) -> list[dict]:
    """
    Call Random User API and return the results list.
    count: number of users to fetch (1–100).
    """
    
    n = max(1, min(100, int(count) if count else 10))
    url = f"{RANDOM_USER_API}?results={n}"
    resp = requests.get(url, timeout=30)
    resp.raise_for_status()
    data = resp.json()
    
    # Fallback users in case the API returns an empty `results` array.
    users = [
    {"first_name": "Liam", "last_name": "Cohen", "age": 28},
    {"first_name": "Emma", "last_name": "Levin", "age": 34},
    {"first_name": "Noah", "last_name": "Goldberg", "age": 22},
    {"first_name": "Ava", "last_name": "Mizrahi", "age": 31},
    {"first_name": "Ethan", "last_name": "Haddad", "age": 26},
    {"first_name": "Mia", "last_name": "Shapira", "age": 29},
    {"first_name": "Daniel", "last_name": "Azoulay", "age": 41},
    {"first_name": "Sophia", "last_name": "Barak", "age": 24},
    {"first_name": "Lucas", "last_name": "Peretz", "age": 37},
    {"first_name": "Ella", "last_name": "Navon", "age": 33},
    ]
    results = data.get("results") if isinstance(data, dict) else None
    if not results:
        # Only override when the API returned no results.
        if isinstance(data, dict):
            data["results"] = users[:n]
        results = users[:n]
    return results or []


class FetchUsersAgent:
    """Node function: uses random_users_api to fetch users, returns partial state update."""

    def __init__(self) -> None:
        self.name = "FetchUsersAgent"

    def run(self, state: dict) -> dict:
        user_count = state.get("user_count", 10)
        users = random_users_api(user_count)
        return {"random_users": users}

    def get_info(self) -> dict:
        return {"name": self.name, "role": "Fetch random users via Random User API"}
